source/common/upstream/load_balancer_impl.h

```diff
@@ -470,6 +476,9 @@ class LeastRequestLoadBalancer : public EdfLoadBalancerBase {
   HostConstSharedPtr unweightedHostPick(const HostVector& hosts_to_use,
                                         const HostsSource& source) override;
   const uint32_t choice_count_;
+
+  // List of hosts per HostsSource for fair random sampling.
+  std::unordered_map<HostsSource, HostVector, HostsSourceHash> unweighted_hosts_;
```

Comment:
I wonder if there is a way to avoid having to allocate a second set of host lists - this change means a memory increase of O(hosts * worker threads) for any cluster using LR, which I imagine would be noticeable for deployments with a large amounts of endpoints. 

Feels like we either need to make this not require this much additional memory (can we shuffle the host lists maintained on the `HostSet` instead? Not clear if we want to make that mutable) or make it possible to fall back to an algorithm that doesn't require tracking this additional state. 

Another option would be to allocate a `HostVector` (or a `std::vector<size_t>` if you're just trying to shuffle the indices around) during load balancing - this would trade additional allocations + vector construction in the LB path for not having to maintain all these lists everywhere.

@jmarantz @mattklein123 for thoughts

Reply:
One idea: what if we use the old implementation/algorithm for large numbers of Hosts, and this for small numbers of hosts? (with a configurable threshold perhaps). IIUC, these changes address issues mainly for smaller numbers of hosts. And with smaller numbers, the extra memory is negligible.

Reply:
`HostVector` is a typedef for `std::vector<HostSharedPtr>`, so ultimately its memory footprint is equivalent to `std::vector<size_t>`, which is 8 extra bytes per host per worker on amd64 architecture.
Allocating this vector on every pick attempt (especially assuming large number of endpoints) likely is not performance-viable from cpu usage point of view.

Reply:
Yeah my suggesting to use a `size_t` vector was based on me thinking copying `std::shared_ptr`s would be more expensive than generating a sequence of number, not as a way to reduce the memory footprint.

Gating the behavior on the number of hosts might give us the best of both worlds, though that of course begs the question of what the threshold should be?

Reply:
Yeah I agree that both copying this on every worker and every update is not optimal. 

sizeof(shared_ptr) is typically bigger than 8. I think it's usually 16 (pointer + pointer to control block) but I'm not sure if the current stdlib implementation. Either way not huge.

Per @snowp I think it would be possible to allocate a vector<uint32_t>, and initialize it with indexes 1..n, and then do the same swapping stuff. This might be slightly more compact but require some indirection back into the primary vector.

Per @ggreenway I agree it would be nice to potentially just turn on this optimization when hosts are below some size N?

Reply:
I can enable it conditionally - just let me know the constant.
Are we talking about hundreds or thousands of endpoints?

Reply:
No huge opinion here. Maybe run some benchmarks? Perhaps > 100? Anyone else have any thoughts on this?

Reply:
In addition to a benchmark, can you run a simulation to record fairness (quality of LB decisions) of the two alternatives at different cluster sizes?

Reply:
Is there a simulator written anywhere in the codebase? There is `DISABLED_SimulationTest` and looks like it hard-codes random load balancer.
Should I just use it with different policies or are there better options?

Reply:
I think we had a benchmark that was working, but maybe the DISABLED one is broken. @tonya11en do you remember the history here? @antoniovicente has also been fixing some of the benchmarks and can help.

Reply:
I recall the simulations always being disabled, but one of them either wouldn't compile or suffered from a runtime error.

Reply:
The number of attempts is usually small.  An alternate approach could be to track indexes picked so far in an array or set and try different random numbers until N different backend indexes are returned.

Another thing to consider is that hitting duplicates may be beneficial.  Take the degenerate case of having only 2 backends and 2 pick attempts.  This will result in the backend with the least connections being picked 3/4 of the time.  If we have a situation where the backend with least requests is in that state because it is actively load shedding or behaving differently than the backends with higher connection counts, picking the backend with the least requests more often results in more requests failing.

Reply:
+1 for the first part of this comment: it would seem simpler to just keep a local set of considered indexes and pick a new random number if you wind up comparing the same ones. I think that would be easier to reason about from a memory perspective. You would have to handle the degenerate case of NumHosts <= LeastRequestCardinality, maybe by just setting a local LeastRequestCardinality to NumHosts-1 or similar.

@antoniovicente second paragraph describes a scenario where IMO a user should avoid using LeastRequests in any case. I guess his argument is that such a choice would be less catastrophic in the current implementation than it would be in the new scenario.

Also FWIW, would recommend using absl::flat_hash_set rather than std::unordered_set for perf and size.